### PR TITLE
Cleanup in VS generator UI code.

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**
Pulling out cleanup code from https://github.com/dotnet/roslyn/pull/74444

### CODE REVIEW

File: src/VisualStudio/Core/Impl/SolutionExplorer/AnalyzerItem/AnalyzerItem.cs [hunk 1]
```diff
@@ -9,83 +9,73 @@
 using Microsoft.VisualStudio.Imaging;
```
Comment (CyrusNajmabadi) [inline src/VisualStudio/Core/Impl/SolutionExplorer/AnalyzerItem/AnalyzerItem.cs:9]: view with whitespace off.

File: src/VisualStudio/Core/Impl/SolutionExplorer/AnalyzerItem/AnalyzerItem.cs [hunk 2]
```diff
@@ -9,83 +9,73 @@
...
 using Microsoft.VisualStudio.Imaging.Interop;
 
-namespace Microsoft.VisualStudio.LanguageServices.Implementation.SolutionExplorer
+namespace Microsoft.VisualStudio.LanguageServices.Implementation.SolutionExplorer;
```
Comment (CyrusNajmabadi) [inline src/VisualStudio/Core/Impl/SolutionExplorer/AnalyzerItem/AnalyzerItem.cs:12]: FSNs and primary constructors.

File: src/VisualStudio/Core/Impl/SolutionExplorer/BaseItem.cs [hunk 3]
```diff
@@ -33,11 +33,11 @@ internal abstract class BaseItem :
     {
         public virtual event PropertyChangedEventHandler PropertyChanged { add { } remove { } }
 
-        private readonly string _name;
+        protected readonly string Name;
```
Comment (CyrusNajmabadi) [inline src/VisualStudio/Core/Impl/SolutionExplorer/BaseItem.cs:36]: part of the equality contract the derived tuypes check.

File: src/VisualStudio/Core/Impl/SolutionExplorer/DiagnosticItem/BaseDiagnosticAndGeneratorItemSource.cs [hunk 4]
```diff
@@ -123,7 +123,7 @@ private BulkObservableCollection<BaseItem> CreateDiagnosticAndGeneratorItems(Pro
 
             collection.AddRange(
                 AnalyzerReference.GetGenerators(language)
-                .Select(g => new SourceGeneratorItem(projectId, g, AnalyzerReference)));
+                .Select(g => new SourceGeneratorItem(projectId, SourceGeneratorIdentity.Create(g, AnalyzerReference), AnalyzerReference.FullPath)));
```
Comment (CyrusNajmabadi) [inline src/VisualStudio/Core/Impl/SolutionExplorer/DiagnosticItem/BaseDiagnosticAndGeneratorItemSource.cs:126]: moved SGItem to be pure data.

File: src/VisualStudio/Core/Impl/SolutionExplorer/DiagnosticItem/CpsDiagnosticItemSource.cs [hunk 5]
```diff
@@ -11,112 +11,116 @@
...
-    {
-        private readonly IVsHierarchyItem _item;
-        private readonly string _projectDirectoryPath;
+    private readonly IVsHierarchyItem _item;
+    private readonly string _projectDirectoryPath;
 
-        /// <summary>
-        /// The analyzer reference that has been found. Once it's been assigned a non-null value, it'll never be assigned null again.
-        /// </summary>
-        private AnalyzerReference? _analyzerReference;
```
Comment (CyrusNajmabadi) [inline src/VisualStudio/Core/Impl/SolutionExplorer/DiagnosticItem/CpsDiagnosticItemSource.cs:24]: moved to base class.

File: src/VisualStudio/Core/Impl/SolutionExplorer/DiagnosticItem/DiagnosticItem.cs [hunk 6]
```diff
@@ -2,91 +2,105 @@
...
+    internal Task<Solution> GetSolutionWithUpdatedAnalyzerConfigSeverityAsync(ReportDiagnostic value, Project project, CancellationToken cancellationToken)
+    {
+        var effectiveSeverity = value.ToDiagnosticSeverity() ?? Descriptor.DefaultSeverity;
+        var diagnostic = Diagnostic.Create(Descriptor, Location.None, effectiveSeverity, additionalLocations: null, properties: null);
+        return ConfigurationUpdater.ConfigureSeverityAsync(value, diagnostic, project, cancellationToken);
+    }
 
-            ruleSetDocument.Save(pathToRuleSet);
-        }
+    public override int GetHashCode()
```
Comment (CyrusNajmabadi) [inline src/VisualStudio/Core/Impl/SolutionExplorer/DiagnosticItem/DiagnosticItem.cs:90]: equals/gethashcode are used in next PR as part of proper items in a observable-collectoin and virtualizing-tree-view.

File: src/VisualStudio/Core/Impl/SolutionExplorer/AnalyzerItem/AnalyzerItem.cs [hunk 7]
```diff
@@ -9,83 +9,73 @@
...
-        public override object GetBrowseObject()
-        {
-            return new BrowseObject(this);
-        }
+    public override object GetBrowseObject()
+    {
+        return new BrowseObject(this);
+    }
 
-        public AnalyzerReference AnalyzerReference
```
Comment (ToddGrun) [inline src/VisualStudio/Core/Impl/SolutionExplorer/AnalyzerItem/AnalyzerItem.cs:56]: >AnalyzerReference [](http://example.com/codeflow?start=33&length=17)

could either of the properties directly below this be changed too?

File: src/VisualStudio/Core/Impl/SolutionExplorer/DiagnosticItem/CpsDiagnosticItemSourceProvider.cs [hunk 8]
```diff
@@ -7,162 +7,162 @@
...
-        public CpsDiagnosticItemSourceProvider(
-            [Import(typeof(AnalyzersCommandHandler))] IAnalyzersCommandHandler commandHandler,
-            IDiagnosticAnalyzerService diagnosticAnalyzerService,
-            VisualStudioWorkspace workspace)
-        {
-            _commandHandler = commandHandler;
-            _diagnosticAnalyzerService = diagnosticAnalyzerService;
-            _workspace = workspace;
-        }
+    private readonly IThreadingContext _threadingContext = threadingContext;
```
Comment (ToddGrun) [inline src/VisualStudio/Core/Impl/SolutionExplorer/DiagnosticItem/CpsDiagnosticItemSourceProvider.cs:38]: >_threadingContext [](http://example.com/codeflow?start=39&length=17)

I assume usage of _threadingContext and _listenerProvider is in the associated PR?

File: src/VisualStudio/Core/Impl/SolutionExplorer/AnalyzerItem/AnalyzerItem.cs (hunk 7 above)
Comment (CyrusNajmabadi) [reply src/VisualStudio/Core/Impl/SolutionExplorer/AnalyzerItem/AnalyzerItem.cs:56]: yes.  done.

File: src/VisualStudio/Core/Impl/SolutionExplorer/DiagnosticItem/DiagnosticItem.cs [hunk 9]
```diff
@@ -2,91 +2,105 @@
 // The .NET Foundation licenses this file to you under the MIT license.
 // See the LICENSE file in the project root for more information.
 
+using System;
+using System.Collections.ObjectModel;
 using System.ComponentModel;
 using System.Threading;
 using System.Threading.Tasks;
 using System.Xml.Linq;
+using ICSharpCode.Decompiler.IL;
```
Comment (ToddGrun) [inline src/VisualStudio/Core/Impl/SolutionExplorer/DiagnosticItem/DiagnosticItem.cs:11]: >sing ICSharpCode.Decompiler.IL; [](http://example.com/codeflow?start=1&length=31)

intentional?

File: src/VisualStudio/Core/Impl/SolutionExplorer/DiagnosticItem/CpsDiagnosticItemSourceProvider.cs (hunk 8 above)
Comment (CyrusNajmabadi) [reply src/VisualStudio/Core/Impl/SolutionExplorer/DiagnosticItem/CpsDiagnosticItemSourceProvider.cs:38]: yes.

File: src/VisualStudio/Core/Impl/SolutionExplorer/DiagnosticItem/DiagnosticItem.cs (hunk 9 above)
Comment (CyrusNajmabadi) [reply src/VisualStudio/Core/Impl/SolutionExplorer/DiagnosticItem/DiagnosticItem.cs]: nope.  removed!

File: src/VisualStudio/Core/Impl/SolutionExplorer/DiagnosticItem/DiagnosticItem.cs [hunk 10]
```diff
@@ -2,91 +2,105 @@
...
-            var effectiveSeverity = value.ToDiagnosticSeverity() ?? Descriptor.DefaultSeverity;
-            var diagnostic = Diagnostic.Create(Descriptor, Location.None, effectiveSeverity, additionalLocations: null, properties: null);
-            return ConfigurationUpdater.ConfigureSeverityAsync(value, diagnostic, project, cancellationToken);
-        }
+    public override bool Equals(object obj)
+        => Equals(obj as DiagnosticItem);
+
+    public bool Equals(DiagnosticItem? other)
+    {
+        return other != null &&
```
Comment (ToddGrun) [inline src/VisualStudio/Core/Impl/SolutionExplorer/DiagnosticItem/DiagnosticItem.cs:98]: >return other != null && [](http://example.com/codeflow?start=8&length=23)

No referenceequality check wanted?

Comment (CyrusNajmabadi) [reply src/VisualStudio/Core/Impl/SolutionExplorer/DiagnosticItem/DiagnosticItem.cs:98]: sure.  i can add that.

Comment (CyrusNajmabadi) [reply src/VisualStudio/Core/Impl/SolutionExplorer/DiagnosticItem/DiagnosticItem.cs:101]: done.

File: src/VisualStudio/Core/Impl/SolutionExplorer/DiagnosticItem/LegacyDiagnosticItemSourceProvider.cs [hunk 11]
```diff
@@ -5,40 +5,33 @@
...
+
+[Export(typeof(IAttachedCollectionSourceProvider))]
+[Name(nameof(LegacyDiagnosticItemSourceProvider))]
+[Order]
+[AppliesToProject("(CSharp | VB) & !CPS")]
+[method: ImportingConstructor]
+[method: Obsolete(MefConstruction.ImportingConstructorMessage, error: true)]
+internal sealed class LegacyDiagnosticItemSourceProvider(
+    [Import(typeof(AnalyzersCommandHandler))] IAnalyzersCommandHandler commandHandler,
+    IDiagnosticAnalyzerService diagnosticAnalyzerService) : AttachedCollectionSourceProvider<AnalyzerItem>
```
Comment (ToddGrun) [inline src/VisualStudio/Core/Impl/SolutionExplorer/DiagnosticItem/LegacyDiagnosticItemSourceProvider.cs:25]: >diagnosticAnalyzerService [](http://example.com/codeflow?start=31&length=25)

What's the rule of thumb that you use to determine whether to declare a field for these?

Comment (CyrusNajmabadi) [reply src/VisualStudio/Core/Impl/SolutionExplorer/DiagnosticItem/LegacyDiagnosticItemSourceProvider.cs:25]: if the file is 'big' :)  

if it's small (just a tiny number of methods, or like one method that then uses the PC params), then i happily use the params.  If it's large, i'm amenable to people who say they find it difficult to reason about, and i opt for fields.

**Review by ToddGrun — APPROVED** [review]
:shipit:

**Comment by CyrusNajmabadi** [comment]
ok.  will ping once this goes in and i meerge this into the other pr.

**Review by jasonmalinowski — APPROVED** [review]

**Pull request merged**